banking_system/maintenance.cpp

```diff
@@ -0,0 +1,74 @@
+#include "maintenance.h"
+
+#include <algorithm>
+#include <iterator>
+#include <iostream>
+#include <stdexcept>
+#include <utility>
+
+#include "account.h"
+#include "premium_account_manager.h"
+#include "standard_account_manager.h"
+#include "exceptions/client_not_found.h"
+#include "exceptions/no_accounts_found_for_given_client_id.h"
+#include "exceptions/no_clients_present.h"
+
+
+namespace bank {
+    Maintenance::Maintenance() {
+        m_clients = std::vector<std::shared_ptr<Person>>();
+        m_accounts = std::vector<std::unique_ptr<IAccount>>();
+    }
+
+    void Maintenance::listClients() const {
+        if (m_clients.empty())
+            throw exceptions::NoClientsPresent();
+
+        for(auto const& client: m_clients) {
+            std::cout << "Client ID: " << client->id << ". Details: " << client->getPersonDetails() << std::endl;
+        }
+    }
+
+    void Maintenance::listClientAccounts(std::vector<std::vector<std::unique_ptr<IAccount>>::iterator> const& accounts) {
+        for(int i = 0; i < accounts.size(); ++i) {
+            std::cout << "Account nr " << i << ". Client: " << (*accounts[i])->getDetails() << std::endl;
+        }
+    }
+
+    std::vector<std::vector<std::unique_ptr<IAccount>>::iterator> Maintenance::findClientAccounts(unsigned int const clientId) {
```

Comment:
why we use here nested vector? Can we use such approach:
```cpp
std::vector<std::unique_ptr<IAccount>>::iterator Maintenance::findClientAccount(unsigned int const clientId) {
    auto isClientAccount = &clientId {
        return clientId == account->getUserId();
    };

    auto it = std::find_if(m_accounts.begin(), m_accounts.end(), isClientAccount);
    if (it == m_accounts.end()) {
        throw exceptions::NoAccountFoundForGivenClientId();
    }

    return it;
}
```

Reply:
As well, I suggest to not return iterator to structure inside class, because when client code will try to use it - it can be change and we will have invalid iterator. Better option - return value, or iterator that can store value inside some how.